### DISCUSSION TITLE
About loss label generate

### DISCUSSION THREAD

**Discussion by hutao568** [fredzzhang/spatio-attentive-graphs#49] [Q&A] [answered]
Hello, I am a bit confused about the calculation of the classification loss in your code. If you have time, can you give me an answer?
https://github.com/fredzzhang/spatio-attentive-graphs/blob/main/interaction_head.py#L182
In your postprocess function, you pass in the one-hot form of the prediction result of each pair and the one-hot form of the label. Next, you use a series of magic operations to convert the one-hot form of the label into one I don't quite understand things.

**Comment by fredzzhang** [comment, marked as answer]
Hi, @hutao568 

I assume you are referring to these lines.

```python
215            # Keep valid classes
216            x, y = torch.nonzero(p[0]).unbind(1)
217
218            result_dict = dict(
219                boxes_h=b_h, boxes_o=b_o,
220                index=x, prediction=y,
221                scores=s[x, y] * p[:, x, y].prod(dim=0) * w[x].detach(),
222                object=o, prior=p[:, x, y], weights=w
223            )
```

So the argument `prior` refers to the product of object detection scores, which have zero values at action classes that are not valid for an object type (e.g. eating a car or something). And `p` here refers to those prior scores in one image. What line `216` is doing is that it only keeps the valid action classes using the knowledge that invalid ones will have values of zeros (see line `463` to `493` to compute the prior scores).

Hope that clears things up.

Cheers,
Fred
